### PR TITLE
Revise perf scalability doc

### PR DESCRIPTION
**Pull request by allenlsy**
Main changes to the perf & scalability docs:

* Additional scenarios tested: OSM in permissive mode and traffic via IngressBackend.
* Data plane resource consumption regarding number of connections.

Resolves https://github.com/openservicemesh/osm/issues/2993 and https://github.com/openservicemesh/osm-perftest/issues/34

Signed-off-by: Allen Leigh <allenlsy@gmail.com>

### CODE REVIEW

File: content/docs/guides/ha_scale/perf-scale.md [hunk 1]
```diff
@@ -108,8 +117,39 @@ There is not much difference between pod-rollout time and stable time in terms o
...
 
-It is also worth mentioning that enabling permissive mode will cause data plane to use much more memory, as each sidecar needs to keep the connection configuration between all other services. The chart below shows the sidecar memory usage when 200 and 400 pods are added in a permissive-mode enabled mesh. Compared to the non-permissive mode, the memory usage is more than doubled. Similar things happen on the control plane.
+### Other Scenarios
+
+#### Permissive mode
+
+When permissive mode is enabled in OSM, services are allowed to communicate with any other services in the mesh without additional traffic policies. This is implemented by adding an allow-all wildcard RBAC policy in the inbound listener and all known services as outbound clusters in the Envoy sidecar. The same load test described above is performed on a service mesh with permissive mode enabled.
+
+On the control plane, the CPU and memory usage both increase compared to non-permissive mode.
```
Comment (zr-msft) [inline content/docs/guides/ha_scale/perf-scale.md:126]: which are you saying:
- CPU and memory usage is higher when running permissive mode?
- CPU and memory usage rates increased at the same rate between permissive and non-permissive mode so their overall usage is the same?

Comment (zr-msft) [inline content/docs/guides/ha_scale/perf-scale.md:126]: which are you saying:
- CPU and memory usage is higher when running permissive mode?
- CPU and memory usage rates increased at the same rate between permissive and non-permissive mode so their overall usage is the same?

File: content/docs/guides/ha_scale/perf-scale.md [hunk 2]
```diff
@@ -108,8 +117,39 @@ There is not much difference between pod-rollout time and stable time in terms o
...
+
+On the data plane, there is no obvious difference in terms of CPU usage. However, the memory usage is much higher in permissive mode. This is expected as more outbound cluster information is stored in the sidecar. The chart below shows the sidecar memory usage when 10 and 20 services are added in permissive mesh and non-permissive mesh respectively. Compared to the non-permissive mode, the memory usage is more than doubled.
 
 <p align="center">
   <img src="/docs/images/perf/perm-memory.png" width="650"/>
 </p>
+
+The request latency from load generator to workloads is lower in permissive mode. This is because that the RBAC policy filter path is shorter, which is a single wildcard filter as described above. There is less rule matching process for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:
```
Comment (zr-msft) [inline content/docs/guides/ha_scale/perf-scale.md:134]:
```suggestion
The request latency from load generator to workloads is lower in permissive mode. This reduction is due to the RBAC policy filter path being a single wildcard filter as described above. This filter path is shorter than non-permissive mode and there is less rule matching operations for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, the larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:
```

Comment (zr-msft) [inline content/docs/guides/ha_scale/perf-scale.md:134]:
```suggestion
The request latency from load generator to workloads is lower in permissive mode. This reduction is due to the RBAC policy filter path being a single wildcard filter as described above. This filter path is shorter than non-permissive mode and there is less rule matching operations for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, the larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:
```

File: content/docs/guides/ha_scale/perf-scale.md [hunk 3]
```diff
@@ -108,8 +117,39 @@ There is not much difference between pod-rollout time and stable time in terms o
...
+
+In this test, the load generator is deployed outside of the mesh. The requests enter the mesh via [IngressBackend](https://release-v1-1.docs.openservicemesh.io/docs/guides/traffic_management/ingress/#ingressbackend-api). Therefore, the request only goes through one Envoy sidecar instead of two.
+
+We compared three scenarios:
+
+* Normal: load generator is inside the mesh. Request goes through 2 sidecars.
+* Ingress: load generator is outside the mesh. Request goes through 1 sidecar.
+* No OSM installed: Request does not go through any sidecar.
+
+The test result shows that, the latency of the ingress scenario drops a lot compared to that in the normal scenario. The difference is roughly equals to the latency of ingress scenario - the latency when OSM is not installed, which is the overhead added by a single Envoy sidecar. This complies with our assumption.
```
Comment (zr-msft) [inline content/docs/guides/ha_scale/perf-scale.md:150]:
```suggestion
The test result shows that the latency of the ingress scenario is significantly lower than the normal scenario. The ingress scenario is roughly equal to the results of the scenario without OSM installed. This result shows that the Envoy sidecar does add latency overhead.
```

Comment (zr-msft) [inline content/docs/guides/ha_scale/perf-scale.md:150]:
```suggestion
The test result shows that the latency of the ingress scenario is significantly lower than the normal scenario. The ingress scenario is roughly equal to the results of the scenario without OSM installed. This result shows that the Envoy sidecar does add latency overhead.
```

File: content/docs/guides/ha_scale/perf-scale.md (hunk 1 above)
Comment (allenlsy) [reply content/docs/guides/ha_scale/perf-scale.md:126]: I don't quite understand the second point. Their overall usage cannot be the same.

I mean, for two same systems except the permissive mode config, the one with permissive mode enabled consumes more CPU and memory than the one disabled it.

**Review by zr-msft — APPROVED** [review]
LGTM!

**Review by steeling — COMMENTED** [review]
I'm again skeptical of these large jumps in latency. 

What are the sidecar limits here? for both CPU and memory?

Then what are your actual VM limits, and how many sidecars do you have operating?

Even with large limits, you will be physically limited by bin packing on VM's.

**Comment by allenlsy** [comment]
@steeling Sidecars definition is added by OSM injector. There is no resource limitation configured. In this case, according to k8s doc, it can use unlimited resources on the node. The node in our tests has 2 vCores. I monitored the resource consumption during the tests. Envoy takes 1 vCore CPU at peak.

File: content/docs/guides/ha_scale/perf-scale.md [hunk 4]
```diff
@@ -108,8 +117,39 @@ There is not much difference between pod-rollout time and stable time in terms o
...
 
-It is also worth mentioning that enabling permissive mode will cause data plane to use much more memory, as each sidecar needs to keep the connection configuration between all other services. The chart below shows the sidecar memory usage when 200 and 400 pods are added in a permissive-mode enabled mesh. Compared to the non-permissive mode, the memory usage is more than doubled. Similar things happen on the control plane.
+### Other Scenarios
+
+#### Permissive mode
+
+When permissive mode is enabled in OSM, services are allowed to communicate with any other services in the mesh without additional traffic policies. This is implemented by adding an allow-all wildcard RBAC policy in the inbound listener and all known services as outbound clusters in the Envoy sidecar. The same load test described above is performed on a service mesh with permissive mode enabled.
```
Comment (steeling) [inline content/docs/guides/ha_scale/perf-scale.md:124]: Add that permissive mode still enables mTLS, and requires the services be within the mesh

**Comment by steeling** [comment]
> @steeling Sidecars definition is added by OSM injector. There is no resource limitation configured. In this case, according to k8s doc, it can use unlimited resources on the node. The node in our tests has 2 vCores. I monitored the resource consumption during the tests. Envoy takes 1 vCore CPU at peak.

And how many nodes vs how many sidecars? If you have 10,000 sidecars, but 2 nodes, you're not going to get true results due to physical limitations.

I don't think we should post the results as is, without running the test with more and larger machines

File: content/docs/guides/ha_scale/perf-scale.md [hunk 5]
```diff
@@ -108,8 +117,39 @@ There is not much difference between pod-rollout time and stable time in terms o
 
 By looking at the collected metrics, in our testing scenarios, resource consumption of data plane sidecars does not change much along with the mesh size. At stable time, the average CPU usage is close to 0 and average memory usage is around 50MB. CPU usage is related to request processing and memory usage is related the size of Envoy configuration stored locally. These two factors do not change with the mesh scale.
 
-It is also worth mentioning that enabling permissive mode will cause data plane to use much more memory, as each sidecar needs to keep the connection configuration between all other services. The chart below shows the sidecar memory usage when 200 and 400 pods are added in a permissive-mode enabled mesh. Compared to the non-permissive mode, the memory usage is more than doubled. Similar things happen on the control plane.
+### Other Scenarios
+
+#### Permissive mode
```
Comment (steeling) [inline content/docs/guides/ha_scale/perf-scale.md:122]: is the image `latency-ingress.png` used for anything?

Comment (allenlsy) [reply content/docs/guides/ha_scale/perf-scale.md:122]: Yes. The last section tests performance in a `IngressBackend` ingress mode.

**Comment by allenlsy** [comment]
@steeling Doc updated with p50 and p99 latency in the RPS section and charts with detailed numbers are removed since the latency between different RPS levels are similar.

**Comment by steeling** [comment]
> > @steeling Sidecars definition is added by OSM injector. There is no resource limitation configured. In this case, according to k8s doc, it can use unlimited resources on the node. The node in our tests has 2 vCores. I monitored the resource consumption during the tests. Envoy takes 1 vCore CPU at peak.
> 
> And how many nodes vs how many sidecars? If you have 10,000 sidecars, but 2 nodes, you're not going to get true results due to physical limitations.
> 
> I don't think we should post the results as is, without running the test with more and larger machines

@allenlsy the latency graphs don't show a good story. Let's think through why this is happening. Can you answer the above questions?

**Comment by steeling** [comment]
> > > > @steeling Sidecars definition is added by OSM injector. There is no resource limitation configured. In this case, according to k8s doc, it can use unlimited resources on the node. The node in our tests has 2 vCores. I monitored the resource consumption during the tests. Envoy takes 1 vCore CPU at peak.
> > > 
> > > 
> > > And how many nodes vs how many sidecars? If you have 10,000 sidecars, but 2 nodes, you're not going to get true results due to physical limitations.
> > > I don't think we should post the results as is, without running the test with more and larger machines
> > 
> > 
> > @allenlsy the latency graphs don't show a good story. Let's think through why this is happening. Can you answer the above questions?
> 
> Currently in the RPS test, we always use one workload pod to handle requests, so that we don't need to scale up load generator. The latest test result shows that latency is relatively stable across all tested RPS levels. The statement in our doc `OSM adds around 0.5ms and 5ms overhead to the 50th percentile and the 99th percentile latency respectively` is still true.
> 
> RPS	P50 latency	P90 latency	P99 latency	Sidecar CPU usage
> 1000	1.5	2	4.7	0.58
> 2000	1.5	2	4.8	1.1
> 4000	1.5	3	4.8	2.2
> 8000	2	4.5	6.27	4.5
> In this test, the vm node has sufficient CPU cores. In the previous tests, our cluster uses nodes with 2 vCores. Thus we saw huge latency spike at RPS 4000. As we can see from the chart above, RPS 4000 requires 2.2 cores of CPU process smoothly. This explains why we hit the performance ceiling before.

Can you please comment on the numbers of machines/cores/sizes? 

From those results it looks like 1.5 ms, not 0.5ms

**Comment by steeling** [comment]
In fact, let's add the test environment that we ran this in into the doc itself. 

ie: what cluster, what number of cores, how many machines, etc.

File: content/docs/guides/ha_scale/perf-scale.md [hunk 6]
```diff
@@ -8,7 +8,7 @@ Open Service Mesh aims to create a high-performance service mesh while providing
 
 ## Testing environment
 
-This test is performed on a Kubernetes cluster provisioned and managed by [Microsoft Azure Kubernetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/). The cluster has 100 Standard_DS2_v2 nodes (each node has 2 vCPUs and 7GB memory). It is ensured that there is only one OSM controller pod running with default resource quota during the test. OSM is configured in non-permissive traffic mode.
+This test is performed on a Kubernetes cluster provisioned and managed by [Microsoft Azure Kubernetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/). The cluster has 100 Standard_DS2_v2 nodes (each node has 2 vCPUs and 7GB memory) and 3 Standard_D32s_v3 nodes (32 vCPUs and 128GB memory) reserved for high intensity tests. It is ensured that there is only one OSM controller pod running with default resource quota during the test. OSM is configured in non-permissive traffic mode.
```
Comment (steeling) [inline content/docs/guides/ha_scale/perf-scale.md:11]: what does 3 nodes reserved for high intensity tests mean?

File: content/docs/guides/ha_scale/perf-scale.md [hunk 7]
```diff
@@ -18,56 +18,24 @@ The testing mesh contains two types of services. One is a load generator (modifi
...
-</p>
+In our test, regardless of the RPS level, OSM adds around 0.5ms and 5ms overhead to the 50th percentile and the 99th percentile latency respectively. This is caused by the request processing in both the client and server sidecars, for tasks like mTLS authentication, metrics collection etc. Our test setup was with a single client and server. Performance may vary with more complex applications, or with different traffic policies set on the mesh.
+
+#### Test 2: Relation with number of connections
+
+A similar setup to RPS scenario is used to test the relationship between latency and number of connections to the sidecar proxy. In this test, the per pod RPS is set at 2500. As the number of connections increased, the request latency showed a proportional increase while the sidecar memory usage to maintain those open connections only slightly increased. The sidecar consumes around 5MB memory for every 100 connections maintained.
 
 <p align="center">
-  <img src="/docs/images/perf/latency-w-osm.png" width="650"/>
+  <img src="/docs/images/perf/latency-conns.png" width="650"/>
```
Comment (steeling) [inline content/docs/guides/ha_scale/perf-scale.md:36]: this graph doesn't tell a great story, can we try this with a bigger node pool?

File: content/docs/guides/ha_scale/perf-scale.md [hunk 8]
```diff
@@ -108,8 +76,39 @@ There is not much difference between pod-rollout time and stable time in terms o
...
 
 <p align="center">
   <img src="/docs/images/perf/perm-memory.png" width="650"/>
 </p>
+
+The request latency from load generator to workloads is lower in permissive mode. This reduction is due to the RBAC policy filter path being a single wildcard filter as described above. This filter path is shorter than non-permissive mode and there is less rule matching operations for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, the larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:
+
+<p align="center">
+  <img src="/docs/images/perf/latency-perm-vs-non-perm.png" width="650"/>
```
Comment (steeling) [inline content/docs/guides/ha_scale/perf-scale.md:96]: similarly these are untenable results. i'm virtually positive that there is a bottleneck unrelated to OSM that is yielding these results.

Why can't we run this on a large cluster?

**Comment by allenlsy** [comment]
> > > > > @steeling Sidecars definition is added by OSM injector. There is no resource limitation configured. In this case, according to k8s doc, it can use unlimited resources on the node. The node in our tests has 2 vCores. I monitored the resource consumption during the tests. Envoy takes 1 vCore CPU at peak.
> > > > 
> > > > 
> > > > And how many nodes vs how many sidecars? If you have 10,000 sidecars, but 2 nodes, you're not going to get true results due to physical limitations.
> > > > I don't think we should post the results as is, without running the test with more and larger machines
> > > 
> > > 
> > > @allenlsy the latency graphs don't show a good story. Let's think through why this is happening. Can you answer the above questions?
> > 
> > 
> > Currently in the RPS test, we always use one workload pod to handle requests, so that we don't need to scale up load generator. The latest test result shows that latency is relatively stable across all tested RPS levels. The statement in our doc `OSM adds around 0.5ms and 5ms overhead to the 50th percentile and the 99th percentile latency respectively` is still true.
> > RPS	P50 latency	P90 latency	P99 latency	Sidecar CPU usage
> > 1000	1.5	2	4.7	0.58
> > 2000	1.5	2	4.8	1.1
> > 4000	1.5	3	4.8	2.2
> > 8000	2	4.5	6.27	4.5
> > In this test, the vm node has sufficient CPU cores. In the previous tests, our cluster uses nodes with 2 vCores. Thus we saw huge latency spike at RPS 4000. As we can see from the chart above, RPS 4000 requires 2.2 cores of CPU process smoothly. This explains why we hit the performance ceiling before.
> 
> Can you please comment on the numbers of machines/cores/sizes?
> 
> From those results it looks like 1.5 ms, not 0.5ms



> > > > > @steeling Sidecars definition is added by OSM injector. There is no resource limitation configured. In this case, according to k8s doc, it can use unlimited resources on the node. The node in our tests has 2 vCores. I monitored the resource consumption during the tests. Envoy takes 1 vCore CPU at peak.
> > > > 
> > > > 
> > > > And how many nodes vs how many sidecars? If you have 10,000 sidecars, but 2 nodes, you're not going to get true results due to physical limitations.
> > > > I don't think we should post the results as is, without running the test with more and larger machines
> > > 
> > > 
> > > @allenlsy the latency graphs don't show a good story. Let's think through why this is happening. Can you answer the above questions?
> > 
> > 
> > Currently in the RPS test, we always use one workload pod to handle requests, so that we don't need to scale up load generator. The latest test result shows that latency is relatively stable across all tested RPS levels. The statement in our doc `OSM adds around 0.5ms and 5ms overhead to the 50th percentile and the 99th percentile latency respectively` is still true.
> > RPS	P50 latency	P90 latency	P99 latency	Sidecar CPU usage
> > 1000	1.5	2	4.7	0.58
> > 2000	1.5	2	4.8	1.1
> > 4000	1.5	3	4.8	2.2
> > 8000	2	4.5	6.27	4.5
> > In this test, the vm node has sufficient CPU cores. In the previous tests, our cluster uses nodes with 2 vCores. Thus we saw huge latency spike at RPS 4000. As we can see from the chart above, RPS 4000 requires 2.2 cores of CPU process smoothly. This explains why we hit the performance ceiling before.
> 
> Can you please comment on the numbers of machines/cores/sizes?
> 
> From those results it looks like 1.5 ms, not 0.5ms

Overhead refers to the extra time consumption compared to the case that no OSM is installed. 0.5ms Overhead is correct in this case, not 1.5ms.

But the numbers will be updated again as we re-run the whole test suite.

File: content/docs/guides/ha_scale/perf-scale.md (hunk 6 above)
Comment (allenlsy) [reply content/docs/guides/ha_scale/perf-scale.md]: This special high intensity nodepool was created and only has 3 nodes due to my resource quota of the cluster.

I've just upscaled my cluster. We will re-run the tests with larger nodes and cluster.

**Review by phillipgibson — APPROVED** [review]

**Comment by shashankram** [comment]
Hey @steeling, since you were reviewing this PR already and had some concerns, could you verify those are resolved? The PR was merged with some unresolved comments, so would be good to confirm all necessary data is captured in the doc.

**Comment by allenlsy** [comment]
Hi @steeling , want to look back to this PR. Can you check again on the comments, and see if we can publish the revised doc?

**Pull request merged**